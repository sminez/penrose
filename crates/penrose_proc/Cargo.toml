[package]
name = "penrose_proc"
version = "0.1.2"
authors = ["IDAM <innes.andersonmorrison@gmail.com>"]
edition = "2018"
license = "MIT"
repository = "https://github.com/sminez/penrose"
documentation = "https://docs.rs/penrose_proc"
readme = "README.md"
description = "proc-macros for the Penrose tiling window manager library"
keywords = [ "windowmanager", "window", "manager", "x11", "procmacro" ]

[lib]
proc-macro = true

[dev-dependencies]
rustversion = "1.0"
trybuild = "1.0"

[dependencies]
syn = { version = "1.0", features = ["fold", "full"] }
quote = "1.0"
proc-macro2 = "1.0"

penrose_keysyms = { version = "0.1.0", path = "../penrose_keysyms" }
strum = "0.19.2"
