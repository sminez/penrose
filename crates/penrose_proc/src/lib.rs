//! Proc macros for use in the main Penrose crate
#![warn(
    clippy::complexity,
    clippy::correctness,
    clippy::style,
    future_incompatible,
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms
)]

use proc_macro::TokenStream;

mod stub;
mod validate_bindings;

use stub::stubbed_companion_trait_inner;
use validate_bindings::validate_user_bindings_inner;

/// This is an internal macro that is used as part of `gen_keybindings` to validate user provided
/// key bindings at compile time using xmodmap.
///
/// It is not intended for use outside of that context and may be modified and updated without
/// announcing breaking API changes.
///
/// ```no_run
/// # use penrose_proc::validate_user_bindings;
/// validate_user_bindings!(
///     ( "M-a", "M-b" )
///     (
///         ( ( "M-{}", "M-S-{}" ) ( "1", "2", "3" ) )
///     )
/// );
/// ```
#[proc_macro]
pub fn validate_user_bindings(input: TokenStream) -> TokenStream {
    validate_user_bindings_inner(input)
}

/// Auto derive a dummy implementation of the marked trait for use in test cases.
///
/// This is _not_ a full blown mocking implemention, it simply allows for quickly setting default
/// return values on trait methods you don't need for a given test case and allows for easily
/// testing functionality that only makes use of a handful of methods.
///
/// Wherever possible, you should always prefer breaking traits down into small, composible units
/// instead of "mega-traits" but in the case of `XConn` and some of its component parts, further
/// granularity isn't really desirable.
#[proc_macro_attribute]
pub fn stubbed_companion_trait(args: TokenStream, input: TokenStream) -> TokenStream {
    stubbed_companion_trait_inner(args, input)
}
