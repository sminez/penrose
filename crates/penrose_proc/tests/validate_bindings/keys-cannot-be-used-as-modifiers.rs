// Bindings with using key names as modifiers are rejected
use penrose_proc::validate_user_bindings;

fn main() {
    validate_user_bindings!(("Return-a")());
}
