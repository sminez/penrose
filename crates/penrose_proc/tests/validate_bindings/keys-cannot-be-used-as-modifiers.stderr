error: proc macro panicked
 --> $DIR/keys-cannot-be-used-as-modifiers.rs:5:5
  |
5 |     validate_user_bindings!(("Return-a")());
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: 'Return-a' is an invalid key binding: 'Return' is an invalid modifer set: valid modifiers are ["A", "M", "S", "C"]
          Key bindings should be of the form <modifiers>-<key name> or <key name> e.g:  M-j, M-S-slash, M-C-Up, XF86AudioMute
